### DISCUSSION TITLE
iterations/gpu don't scale when using custom sampler

### DISCUSSION THREAD

**Discussion by bartmch** [Lightning-AI/pytorch-lightning#3716] [DDP / multi-GPU / multi-node] [answered]
Hi! I'm currently using Pytorch's weighted random sampler for my multi-class skewed dataset and I've put "use_ddp_sampler" to False.
Everything works well with the custom sampler but it's taking a significantly longer time to train my model on multiple GPUs.
I notice that with the default sampler my 2000 iterations scale well over 4 GPUs, 500 iterations/gpu but with the custom sampler, it doesn't scale and stays at 2000 iterations/GPU over 4 GPUs - highly likely explaining the longer training times.

How can I speed up my training procedure using a custom sampler?

Edit: Trainer settings:
```
trainer = pl.Trainer(   gpus                = args.gpus, 
                            num_nodes           = 1,
                            distributed_backend = 'ddp', 
                            max_epochs          = args.epochs,
                            weights_save_path   = args.weights_save_path, 
                            logger              = True,
                            replace_sampler_ddp = False,
                            checkpoint_callback = checkpoint_callback)
```

**Comment by rohitgr7** [comment]
[WeightedRandomSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.WeightedRandomSampler) is not a distributed sampler right?

**Comment by ananyahjha93** [comment]
@bartmch I think @rohitgr7 is right here, the base class for both DistributedSampler and WeightedRandomSampler is Sampler but its the DistributedSampler which implements the DDP capabilities.

**Comment by bartmch** [comment]
Thanks for your replies! Two silly questions: 1) how can I implement a sampler in Lightning which is ddp and takes care of my unbalanced dataset? 2) weighted sampler is simply giving a sampling probability to each number right? Why doesn't it simply scale on multiple goud? Is it not implemented (yet) or am I understanding it wrong?

**Comment by rohitgr7** [comment, marked as answer]
All you need is a custom sampler here + set `replace_sampler_ddp=False` in `Trainer`. `WeightedRandomSampler` just uses the `weights` which you need to define to sample the batch from the dataset something similar to what boosting algorithms do while sampling. In your use case, you need some kind of `DistributedBalancedSampler` that can do either oversampling or undersampling. There are some discussions [here](https://github.com/pytorch/pytorch/issues/23430) which partially might solve your use-case in the future. But for now neither lightning nor PyTorch has this yet. But you check some custom sampler [here](https://catalyst-team.github.io/catalyst/api/data.html#samplers) that might help.

**Comment by rohitgr7** [comment]
@bartmch I hope the above suggestion might help.

**Comment by bartmch** [comment]
Hey @rohitgr7 apologies for the delay - I'll try the solution this week.

**Reply by bw4sz** [reply]
Any updates here, this seems like a critical need, since the vast majority of classification tasks are unbalanced. What is the suggested go to right now? Write a custom sampler and wrap it in distributed, as in above?

**Comment by gianscarpe** [comment]
Hi! Just found this discussion, seems a critical feature to implement directly in lightning! I'm working on my custom implementation for the moment

@Borda @awaelchli
